kv: update the timestamp of a replayed write's intent

miraradeva opened the pull request:
Previously, if a write was replayed at a higher timestamp(e.g. because
the transaction's write timestamp was pushed), the replayed write would
not update the intent's timestamp as long as the epoch and sequence
numbers were the same. This was ok because when the transaction
committed the intent would be updated to match the transaction's commit
timestamp. This treatment of the timestamp, however, was not consistent
with the lock table, where the replayed write would result in an
increased timestamp of the write's replicated lock. This discrepancy
(i.e. a lock table lock with a higher timestamp than the stored intent)
could cause an infinite lock discovery loop, where a request first
checks the lock table, does not find a conflict (because of the higher
lock table timestamp), tries to evaluate, and encounters the
conflicting intent (at a lower timestamp).

In this patch, we match the lock table behavior and let the replayed
write increase the timestamp of the stored intent. The process is
identical to resolving an intent when a pending transaction's timestamp
is pushed forward and the stored intents of the transaction are updated
to match the new timestamp.

Fixes: https://github.com/cockroachdb/cockroach/issues/112409
Part of: https://github.com/cockroachdb/cockroach/issues/111352

Release note (bug fix): no longer possible to have a infinite lock
re-discovery loop when the lock table has a replicated lock with a
higher timestamp than the corresponding stored intent.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/112410)
<!-- Reviewable:end -->

miraradeva committed:
kv: update the timestamp of a replayed write's intent

Previously, if a write was replayed at a higher timestamp(e.g. because
the transaction's write timestamp was pushed), the replayed write would
not update the intent's timestamp as long as the epoch and sequence
numbers were the same. This was ok because when the transaction
committed the intent would be updated to match the transaction's commit
timestamp. This treatment of the timestamp, however, was not consistent
with the lock table, where the replayed write would result in an
increased timestamp of the write's replicated lock. This discrepancy
(i.e. a lock table lock with a higher timestamp than the stored intent)
could cause an infinite lock discovery loop, where a request first
checks the lock table, does not find a conflict (because of the higher
lock table timestamp), tries to evaluate, and encounters the
conflicting intent (at a lower timestamp).

In this patch, we match the lock table behavior and let the replayed
write increase the timestamp of the stored intent. The process is
identical to resolving an intent when a pending transaction's timestamp
is pushed forward and the stored intents of the transaction are updated
to match the new timestamp.

Fixes: #112409
Part of: #111352

Release note (bug fix): no longer possible to have a infinite lock
re-discovery loop when the lock table has a replicated lock with a
higher timestamp than the corresponding stored intent.

miraradeva reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112410)* status: :shipit: complete! 0 of 0 LGTMs obtained

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/112410#-NhSPcR--NSulOQgLmPw:-NhSPcR0AsPC7PM7lyjh:br6kjkz) (no related file):*
The new test in `replica_test.go` hangs prior to the fix, reproducing the infinite lock re-discovery loop. 

I'm not 100% sure of the implications of calling `mvccResolveWriteIntent`. In particular the changes highlighted by the `ambiguous_writes` test around the `localTs` and the `txnDidNotUpdateMeta` fields.

I listed a few alternative solutions (see #112409) that I'd be happy to try out too.

___


<!-- Sent from Reviewable.io -->

miraradeva commented:
Discussed with Nathan and Arul and concluded this is a risky and probably unnecessary change. We'll pursue one of the other options in #112409.
